firefox-ios/Client/Frontend/Browser/WebView/PullRefreshView.swift

```diff
@@ -0,0 +1,323 @@
+            progressView.heightAnchor.constraint(equalToConstant: UX.progressViewSize * shrinkFactor),
+            progressView.widthAnchor.constraint(equalToConstant: UX.progressViewSize * shrinkFactor)
+        ])
+    }
+
+    /// Starts observing the `UIScrollView` content offset so the refresh can react to content change and show
+    /// the correct state of the pull to refresh view.
+    func startObservingContentScroll() {
+        obeserveTicket = scrollView?.observe(\.contentOffset) { [weak self] _, _ in
+            guard let scrollView = self?.scrollView, scrollView.isDragging
+            else {
+                guard let refreshHasFocus = self?.refreshIconHasFocus, refreshHasFocus else { return }
+                self?.refreshIconHasFocus = false
+                self?.easterEggGif?.removeFromSuperview()
+                self?.easterEggGif = nil
+                self?.obeserveTicket?.invalidate()
+                self?.triggerReloadAnimation()
+                return
+            }
+            let threshold = (self?.computeShrinkingFactor() ?? 1.0) * UX.blinkProgressViewStandardThreshold
+            if scrollView.contentOffset.y < -threshold {
+                self?.blinkBackgroundProgressViewIfNeeded()
+                DispatchQueue.main.asyncAfter(deadline: .now() + UX.easterEggDelayInSeconds) {
+                    self?.showEasterEgg()
+                }
+            } else if scrollView.contentOffset.y != 0.0 {
+                // This check prevents progressView re blink when scrolling the pull refresh before the web view is loaded
+                self?.restoreBackgroundProgressViewIfNeeded()
+                let rotationAngle = -(scrollView.contentOffset.y / (self?.frame.height ?? 0.0)) * .pi * 2
+                UIView.animate(withDuration: 0.1) {
+                    self?.progressView.transform = CGAffineTransform(rotationAngle: rotationAngle * 2.0)
+                }
+            }
+        }
+    }
+
+    private func triggerReloadAnimation() {
+        UIView.animate(withDuration: 0.1,
+                       delay: 0,
+                       options: .curveEaseOut,
+                       animations: {
+            self.progressContainerView.transform = UX.progressViewAnimatedBackgroundFinalAnimationTransform
+        }, completion: { _ in
+            self.progressContainerView.backgroundColor = .clear
+            self.progressContainerView.transform = .identity
+            self.progressView.transform = .identity
+            self.onRefreshCallback()
+        })
+    }
+
+    private func blinkBackgroundProgressViewIfNeeded() {
+        guard !refreshIconHasFocus else { return }
+        refreshIconHasFocus = true
+        let shrinkFactor = computeShrinkingFactor()
+        UIView.animate(withDuration: 0.3,
+                       delay: 0,
+                       usingSpringWithDamping: 0.6,
+                       initialSpringVelocity: 10,
+                       animations: {
```

Comment:
Should these constants go up in the UX struct? 🤔

Reply:
Yes definitely i'll move them there